pkg/policyfilter/rthooks/rthooks.go

```diff
@@ -92,15 +93,14 @@ func createContainerHook(_ context.Context, arg *rthooks.CreateContainerArg) err
 		return err
 	}
 
-	var containerFound bool
-	var container *corev1.ContainerStatus
 	namespace := pod.ObjectMeta.Namespace
-	pod, container, containerFound = arg.Watcher.FindContainer(containerID)
-	if !containerFound {
-		log.WithError(err).Warnf("failed to find container information %s, aborting hook.", containerID)
-	}
 
-	containerName := container.Name
+	containerName := arg.Req.ContainerName
+	if containerName == "" {
+		err := fmt.Errorf("failed to find container information %s, aborting hook", containerID)
+		log.Warn(err)
+		return err
```

Comment:
Not sure if failing the is the right thing to do here. 
If we do not get a containerName, we can still do policies that do not depend on container name.

 I think it's better to i) issue a warning and ii) increase a metric if the container name is not there. This would require adding a new policyfilter metric (see policyfiltermetrics), but this should be relatively simple.
 
 WDYT?

Reply:
Hello!
i) You are totally right, it would be bad to break other filtering functionality if we don't get the container name, I will only leave the warning, thank you!
ii) I also totally agree, but I have a question about a separate metric. I see we have the `error_type` label in `policyfilter_metrics_total` and currently it's not used. What if we set it to `nil` by default as it is now, but in case of an error, we could specify an error string type, for example, `container-name-not-found` or maybe something else?:)
If you think we need to fully separate these two metrics, how do you see the new metric?
For example, I'm thinking about a generic `policyfilter_metrics_error_total` and it would contain `op` and `error_type`.
What would be your preferred solution?:) In the meantime, I will test adding an error message to `error_type` just to understand if it's feasible:)

Reply:
> i) You are totally right, it would be bad to break other filtering functionality if we don't get the container name, I will only leave the warning, thank you!

Great, thanks!

> i) I also totally agree, but I have a question about a separate metric. I see we have the error_type label in policyfilter_metrics_total and currently it's not used. What if we set it to nil by default as it is now, but in case of an error, we could specify an error string type, for example, container-name-not-found or maybe something else?:)

I guess the error label was removed recently: https://github.com/cilium/tetragon/commit/425f91e2be6b83c4b654062d64ec49ec5e072d4f. I'll sync with @lambdanis and figure out what's the best way moving forward (https://github.com/cilium/tetragon/pull/2162#discussion_r1546398091), but I don't think we need to block this PR. We can do the metric in a followup PR.

Thanks!

Reply:
From the logical point of view, it's likely better to have a separate metric as you suggested because if we mix these error types that are not errors in code, this will be very confusing :smiley:

Reply:
Looks like I wrote my latest message right after yours, so it may sound strange if you read it in chronological order, sorry for that:)

> I guess the error label was removed recently: https://github.com/cilium/tetragon/commit/425f91e2be6b83c4b654062d64ec49ec5e072d4f. I'll sync with @lambdanis and figure out what's the best way moving forward (https://github.com/cilium/tetragon/pull/2162#discussion_r1546398091), but I don't think we need to block this PR. We can do the metric in a followup PR.

Thank you! For now, I pushed the change not to abort the hook when we don't find the name (squashed with the previous commit), I also updated the commit message to reflect it there too:)

Reply:
Thank you!
One concern I have in this case is let's say we set the `error` label which should indicate that the operation failed; however, in reality, it's only a warning and we don't abort the hook, so it's a kind of success :sweat_smile: (I hope not to confuse anyone)

Reply:
> let's say we set the error label which should indicate that the operation failed; however, in reality, it's only a warning and we don't abort the hook, so it's a kind of success 

This would be confusing indeed, thanks for pointing this out. I'd expect the metric semantics to be:
`sum(tetragon_policyfilter_operations_total)` - number of attempted operations
`sum(tetragon_policyfilter_operations_total{error=""})` - number of successful operations
`sum(tetragon_policyfilter_operations_total{error!=""})` - number of unsuccessful operations

To count missed container names (which still results in a successful operation), a separate metric seems like a better solution.

Reply:
A separate metric only for error makes sense to me.
An alternative approach would be to have the operation be "get_container_name" but then we would have to count the successes as well.
Either approach is fine by me.

Reply:
Thank you for your thoughts!
I spent some time thinking about an optimal approach and at some point, I realized, that adding new metrics may not be an optimal solution, let me explain:
- Adding a new label specific to this case sounds simple but I think at some point there will be more of such exceptions and adding a new label for each could be too much (but I could be wrong, please let me know)
- Adding a separate label for errors/warnings also sounds great but we need to attach the error/warning to some operations, e.g. add container, delete pod, etc. This means we will end up with the same labels (subsystem, operation + new labels); therefore, I think adding a new generic label could be suboptimal (again, I can be wrong, but I wonder what would be the good design).

Therefore, I decided to add a new `warning` label to `policyfilter_metrics_total`. I implemented it in the latest patch. It's worth noting that there can be multiple warnings that we want to report from the same function, and also if there is a warning, we cannot simply say the operation was successful. I handled it using conditions, so if we report a warning, we don't report success but at the same time, we can report multiple warnings if needed.

Also, after adding the `warning` label, I found a missing return statement [here](https://github.com/cilium/tetragon/blob/main/pkg/policyfilter/rthooks/rthooks.go#L104) and realized there will be no metric at all because we return an error. Therefore, I decided to add the `error` label for such cases. @lambdanis had a concern about high cardinality; therefore, I decided to make errors expected, i.e. not use the error message but have an enum just like with operations and warnings.
In the end, I ended up with these labels
```
tetragon_policyfilter_metrics_total{error="",op="add",subsys="pod-handlers",warning=""} 14
tetragon_policyfilter_metrics_total{error="",op="add-container",subsys="rthooks",warning=""} 4
tetragon_policyfilter_metrics_total{error="",op="add-container",subsys="rthooks",warning="container-name-not-found"} 0
tetragon_policyfilter_metrics_total{error="",op="delete",subsys="pod-handlers",warning=""} 2
tetragon_policyfilter_metrics_total{error="",op="update",subsys="pod-handlers",warning=""} 57
tetragon_policyfilter_metrics_total{error="add-pod-container-failed",op="add-container",subsys="rthooks",warning=""} 0
tetragon_policyfilter_metrics_total{error="add-pod-error",op="add",subsys="pod-handlers",warning=""} 0
tetragon_policyfilter_metrics_total{error="delete-pod-error",op="delete",subsys="pod-handlers",warning=""} 0
tetragon_policyfilter_metrics_total{error="update-pod-error",op="update",subsys="pod-handlers",warning=""} 0
```
These changes can be reviewed in this commit https://github.com/cilium/tetragon/pull/2231/commits/9a695c336fbe2b8919f47788414deb2bfbca26f7

WDYT? :smiley: 

----
P.S. There was a failed CI action related to a duplicate signature, I fixed it.

Reply:
@BonySmoke I guess it's the second time we try and solve the same thing: https://github.com/cilium/tetragon/pull/2282 😅 

Not sure how I feel about the idea of adding an "error" and "warning". I feel it's somewhat confusing.

I would still go with the idea of either having a separate metric for errors or adding a new operation for `get-container-name`. I think a separate metric is probably the simplest approach.


> Also, after adding the warning label, I found a missing return statement [here](https://github.com/cilium/tetragon/blob/main/pkg/policyfilter/rthooks/rthooks.go#L104) and realized there will be no metric at all because we return an error.

Nice catch! From a quick look, this seems to be indeed a bug.

Reply:
My worry with the "warning" label is that it makes it unclear what the metric is actually measuring.

`sum(tetragon_policyfilter_metrics_total)` should be meaningful, and my understanding is that it's the number of attempted operations (there can be bugs, but this would be the end goal at least). So it should be incremented exactly once for each operation - either with empty or specific error. Now, if we have multiple "warnings" for one operations and increment the metric for each of these warnings, then `sum(tetragon_policyfilter_metrics_total)` becomes meaningless. So to count warnings (meaning something went wrong, but the operation completed in the end) a separate metric is a way to go IMO.

Side note: I'm reasoning here about how it *should* look like in my opinion, not how it looks like now. There might be more bugs in how this metric is used, if that's the case they should be fixed :) But it's out of scope for this PR. Thanks for digging into it!

Reply:
I added a very simple metric in https://github.com/cilium/tetragon/pull/2231/commits/85b8663f7f2e9ee7c4e24ae564a5c192eb86b684, hopefully, it's intuitive enough:)
```
# HELP tetragon_policyfilter_hook_container_name_missing_total The total number of operations when the container name was missing in the OCI hook
# TYPE tetragon_policyfilter_hook_container_name_missing_total counter
tetragon_policyfilter_hook_container_name_missing_total 0
```
Wdyt?:)
For now, I didn't add any documentation, just want to know your opinion. Also, I didn't add any labels to keep it as simple as possible but I can add any labels if needed.